ui/puz/src/view/history.ts

```diff
@@ -13,41 +13,40 @@ const slowPuzzleIds = (ctrl: PuzCtrl): Set<string> | undefined => {
   return new Set(ctrl.run.history.filter(r => r.millis > threshold).map(r => r.puzzle.id));
 };
 
+function buttonMaker(vmFilter: boolean, title: string, togglefn: any): VNode {
+  return h(
+    'button.puz-history__filter.button',
+    {
+      class: {
+        active: vmFilter,
+        'button-empty': !vmFilter,
+      },
+      hook: onInsert(e => e.addEventListener('click', togglefn)),
+    },
+    title
+  );
+}
+
 export default (ctrl: PuzCtrl): VNode => {
   const slowIds = slowPuzzleIds(ctrl);
   const noarg = ctrl.trans.noarg;
+  const buttons: VNode[] = [
+    buttonMaker(ctrl.vm.filterFailed, noarg('failedPuzzles'), ctrl.toggleFilterFailed),
+    buttonMaker(ctrl.vm.filterSlow, noarg('slowPuzzles'), ctrl.toggleFilterSlow),
+  ];
+  if (ctrl.vm?.filterSkip !== undefined)
```

Comment:
according to the types `filterSkip` should never be `undefined`

Reply:
Forgive me if I've got this wrong but,

- Storm also makes use of this function and does not make use of skips.
- Therefore `PuzCtrl` has to treat `filterSkip` and `toggleFilterSkip()` as optional to keep compatibility with both `RacerCtrl` and `StormCtrl`

Perhaps it would be clearer to explicitly specify `StormVm` ,`RacerVm` extends `PuzVm` and `StormCtrl`, `RacerCtrl` `implements PuzCtrl`?

Reply:
How about making `renderHistory` polymorphic to circumvent the issue of Storm not using skips?